core/server/worker/src/main/java/alluxio/worker/netty/DataServerReadHandler.java

```diff
@@ -39,88 +38,133 @@
 /**
  * This class handles {@link alluxio.proto.dataserver.Protocol.ReadRequest}s.
  *
- * Protocol: Check {@link alluxio.client.block.stream.NettyPacketReader} for more information.
+ * Protocol: Check {@link alluxio.client.block.stream.NettyPacketReader} for additional information.
  * 1. Once a read request is received, the handler creates a {@link PacketReader} which reads
  *    packets from the block worker and pushes them to the buffer.
  * 2. The {@link PacketReader} pauses if there are too many packets in flight, and resumes if there
  *    is room available.
  * 3. The channel is closed if there is any exception during the packet read/write.
+ *
+ * Threading model:
+ * Only two threads are involved at a given point of time: netty I/O thread, packet reader thread.
+ * 1. The netty I/O thread accepts the read request, handles write callbacks. If any exception
+ *    occurs (e.g. failed to read from netty or write to netty) or the read request is cancelled by
+ *    the client, the netty I/O thread notifies the packet reader thread.
+ * 2. The packet reader thread keeps reading from the file and writes to netty. Before reading a
```

Comment:
Yes, it is 'before' reading from the file (e.g. local block file, ufs file), not from netty. These notifications (cancel, error, eof) can be considered as async packets injected to the underlying file. We can also check these after reading the packet but that complicates the logic unnecessarily. 

Note that we won't read anything after  eof is reached, at least one more packet can be sent after cancel is received, at least one more packet can be attempted (it can fail) to be sent after an error.